### PR TITLE
DH: Make padding always on when X9.42 KDF is used

### LINKED ISSUE #17834: DH derive with X9.42 KDF: provider implementation vs old implementation
I think I have found a difference in the DH key derivation with X9.42 KDF between the provider implementation and the old implementation.

First of all RFC 2631 states under **2.1.2.  Generation of Keying Material**:

_X9.42 provides an algorithm for generating an essentially arbitrary amount of keying material from ZZ. Our algorithm is derived from that algorithm by mandating some optional fields and omitting others.
     KM = H ( ZZ || OtherInfo)
H is the message digest function SHA-1 [FIPS-180] ZZ is the shared secret value computed in Section 2.1.1. **Leading zeros MUST be preserved, so that ZZ occupies as many octets as p**. For instance, if p is 1024 bits, ZZ should be 128 bytes long._

When I look at the old implementation in `crypto/dh/dh_pmeth.c` function `pkey_dh_derive()` it uses `DH_compute_key_padded()` to compute the padded derived secret before applying the KDF.

```
    else if (dctx->kdf_type == EVP_PKEY_DH_KDF_X9_42) {

        unsigned char *Z = NULL;
        size_t Zlen = 0;
        if (!dctx->kdf_outlen || !dctx->kdf_oid)
            return 0;
        if (key == NULL) {
            *keylen = dctx->kdf_outlen;
            return 1;
        }
        if (*keylen != dctx->kdf_outlen)
            return 0;
        ret = 0;
        if ((Zlen = DH_size(dh)) <= 0)
            return 0;
        if ((Z = OPENSSL_malloc(Zlen)) == NULL) {
            ERR_raise(ERR_LIB_DH, ERR_R_MALLOC_FAILURE);
            return 0;
        }
        if (DH_compute_key_padded(Z, dhpubbn, dh) <= 0)   <------------------------- OK
            goto err;
        if (!DH_KDF_X9_42(key, *keylen, Z, Zlen, dctx->kdf_oid,
                          dctx->kdf_ukm, dctx->kdf_ukmlen, dctx->kdf_md))
            goto err;
        *keylen = dctx->kdf_outlen;
        ret = 1;
 err:
        OPENSSL_clear_free(Z, Zlen);
        return ret;
    }
```

However, in the provider implementation in `providers/implementations/exchange/dh_exch.c` function `dh_X9_42_kdf_derive()` it calls `dh_plain_derive()` which itself depends on `pdhctx->pad` to perform a padded derivation or not. 

```
static int dh_X9_42_kdf_derive(void *vpdhctx, unsigned char *secret,
                               size_t *secretlen, size_t outlen)
{
    PROV_DH_CTX *pdhctx = (PROV_DH_CTX *)vpdhctx;
    unsigned char *stmp = NULL;
    size_t stmplen;
    int ret = 0;

    if (secret == NULL) {
        *secretlen = pdhctx->kdf_outlen;
        return 1;
    }

    if (pdhctx->kdf_outlen > outlen) {
        ERR_raise(ERR_LIB_PROV, PROV_R_OUTPUT_BUFFER_TOO_SMALL);
        return 0;
    }
    if (!dh_plain_derive(pdhctx, NULL, &stmplen, 0))
        return 0;
    if ((stmp = OPENSSL_secure_malloc(stmplen)) == NULL) {
        ERR_raise(ERR_LIB_PROV, ERR_R_MALLOC_FAILURE);
        return 0;
    }
    if (!dh_plain_derive(pdhctx, stmp, &stmplen, stmplen))   <---- padding depends on pdhctx->pad which is false by default
        goto err;

    /* Do KDF stuff */
    if (pdhctx->kdf_type == PROV_DH_KDF_X9_42_ASN1) {
        if (!ossl_dh_kdf_X9_42_asn1(secret, pdhctx->kdf_outlen,
                                    stmp, stmplen,
                                    pdhctx->kdf_cekalg,
                                    pdhctx->kdf_ukm,
                                    pdhctx->kdf_ukmlen,
                                    pdhctx->kdf_md,
                                    pdhctx->libctx, NULL))
            goto err;
    }
    *secretlen = pdhctx->kdf_outlen;
    ret = 1;
err:
    OPENSSL_secure_clear_free(stmp, stmplen);
    return ret;
}
```

```
dh_plain_derive():
    if (pdhctx->pad)
        ret = DH_compute_key_padded(secret, pub_key, pdhctx->dh);
    else
        ret = DH_compute_key(secret, pub_key, pdhctx->dh);
```

So in my opinion the provider implementation calculates a wrong secret in case the plain derived secret contains leading zeros, because it uses the un-padded secret for applying the KDF, not the padded secret as the RFC requires.

To fix this, `dh_X9_42_kdf_derive()`  should set `pdhctx->pad = 1` before calling `dh_plain_derive()` to ensure the padded secret is used to apply the KDF to.

### PR DESCRIPTION
**Pull request by t8m**
Fixes #17834

##### Checklist

- [x] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

**Comment by t8m** [comment]
Still working on a test.

File: test/recipes/30-test_evp_data/evppkey_ffdhe.txt [hunk 1]
```diff
@@ -93,6 +93,22 @@ PeerKey=ffdhe2048-2-pub
 Ctrl = dh_pad:1
 SharedSecret=00006620DD85B56EE8540C8040CAC46B7385344A164E4DBDF521F7D99F88FA68EDD295A45E36E0BBD5FF5DE84598824E2CA52ED82ACA918CAECC6B22846D0FC6F0203E8B6963964D11E9E704F83AF1D60E9B1931139E9E9967C4665A831A75D99359A8BA80DD5921E74379AF4CA8DB453EDBC5E669AB17A5254CA6C96794CD5196BE90AF37742C8F6812515FFCC45B08F4158EFF9559F1AEF3665B3D91519DCBC6DF22CD6DA521B86613558602E73D2CA4666972F7D2CB6B46299B1DF2DA29A2A2D99D105E10CB553D6738A9B1DB2A0314C3CF30642D5C44695623D8B95C4426BEA830FB51816B4F086945E9B12A445F42DD68610E3F378A6E69A383D13D85BF
 
+Derive=ffdhe2048-1
```
Comment (slontis) [inline test/recipes/30-test_evp_data/evppkey_ffdhe.txt:96]: Maybe a comment for what this is testing? Are these just arbitrary values used here?

**Review by slontis — APPROVED** [review]

File: doc/man7/EVP_KEYEXCH-DH.pod [hunk 2]
```diff
@@ -15,20 +15,40 @@ Key exchange support for the B<DH> key type.
...
 
+Sets the padding mode for the associated key exchange ctx.
+Setting a value of 1 will turn padding on.
+Setting a value of 0 will turn padding off.
+If padding is off then the derived shared secret may be smaller than the largest
+possible secret size.
+If padding is on then the derived shared secret will have its first bytes filled
+with 0s where necessary to make the shared secret the same size as the largest
+possible secret size. The padding mode parameter is ignored (and padding
+implicitly enabled) in case the KDF type is set to "X942KDF-ASN1"
```
Comment (paulidale) [inline doc/man7/EVP_KEYEXCH-DH.pod:26]: s/in case/when/

File: doc/man7/EVP_KEYEXCH-DH.pod [hunk 3]
```diff
@@ -15,20 +15,40 @@ Key exchange support for the B<DH> key type.
 
 =item "pad" (B<OSSL_EXCHANGE_PARAM_PAD>) <unsigned integer>
 
+Sets the padding mode for the associated key exchange ctx.
+Setting a value of 1 will turn padding on.
+Setting a value of 0 will turn padding off.
+If padding is off then the derived shared secret may be smaller than the largest
+possible secret size.
+If padding is on then the derived shared secret will have its first bytes filled
+with 0s where necessary to make the shared secret the same size as the largest
```
Comment (paulidale) [inline doc/man7/EVP_KEYEXCH-DH.pod:24]: s/0s/zeros/

File: test/recipes/30-test_evp_data/evppkey_ffdhe.txt (hunk 1 above)
Comment (paulidale) [reply test/recipes/30-test_evp_data/evppkey_ffdhe.txt:96]: I agree, there should be a comment here.

Do we know that one of these cases hit the leading zero problem?

**Review by paulidale — COMMENTED** [review]
Two language nits.

File: test/recipes/30-test_evp_data/evppkey_ffdhe.txt (hunk 1 above)
Comment (t8m) [reply test/recipes/30-test_evp_data/evppkey_ffdhe.txt:96]: > Maybe a comment for what this is testing? Are these just arbitrary values used here?

I'll add a comment. Those keys are commented above that they create a shared secret with leading zeros.

Also I've tested that with the fix reverted this testcase fails.

**Comment by t8m** [comment]
@slontis Still OK?

**Comment by t8m** [comment]
ping for second review

**Review by beldmit — APPROVED** [review]
LGTM

**Comment by t8m** [comment]
Merged to master and 3.0 branches. Thank you for the reviews!

**Pull request closed without merging**